pkg/messaging/mqtt/pubsub_test.go

```diff
@@ -0,0 +1,427 @@
+			topic:    topic,
+			clientID: "",
+			err:      mqtt_pubsub.ErrEmptyID,
+			handler:  handler{false, "", msgChan},
+		},
+	}
+	for _, tc := range cases {
+		err = pubsub.Subscribe(tc.clientID, tc.topic, tc.handler)
+		assert.Equal(t, err, tc.err, fmt.Sprintf("%s: expected: %s, but got: %s", tc.desc, err, tc.err))
+
+		if tc.err == nil {
+			expectedMsg := messaging.Message{
+				Publisher: "clientID1",
+				Channel:   channel,
+				Subtopic:  subtopic,
+				Payload:   data,
+			}
+			data, err := proto.Marshal(&expectedMsg)
+			assert.Nil(t, err, fmt.Sprintf("got unexpected error: %s", err))
+
+			token := client.Publish(tc.topic, qos, false, data)
+			token.WaitTimeout(500)
+			assert.Nil(t, token.Error(), fmt.Sprintf("got unexpected error: %s", token.Error()))
+
+			receivedMsg := <-msgChan
+			assert.Equal(t, expectedMsg.Payload, receivedMsg.Payload, fmt.Sprintf("%s: expected %+v got %+v\n", tc.desc, expectedMsg, receivedMsg))
+		}
+	}
+}
+
+func TestPubSub(t *testing.T) {
+	msgChan := make(chan messaging.Message)
+
+	cases := []struct {
+		desc     string
+		topic    string
+		clientID string
+		err      error
+		handler  messaging.MessageHandler
+	}{
+		{
+			desc:     "Subscribe to a topic with an ID",
+			topic:    topic,
+			clientID: "clientid7",
+			err:      nil,
+			handler:  handler{false, "clientid7", msgChan},
+		},
+		{
+			desc:     "Subscribe to the same topic with a different ID",
+			topic:    topic,
+			clientID: "clientid8",
+			err:      nil,
+			handler:  handler{false, "clientid8", msgChan},
+		},
+		{
+			desc:     "Subscribe to a topic with a subtopic with an ID",
+			topic:    fmt.Sprintf("%s.%s", topic, subtopic),
+			clientID: "clientid7",
+			err:      nil,
+			handler:  handler{false, "clientid7", msgChan},
+		},
+		{
+			desc:     "Subscribe to an empty topic with an ID",
+			topic:    "",
+			clientID: "clientid7",
+			err:      mqtt_pubsub.ErrEmptyTopic,
+			handler:  handler{false, "clientid7", msgChan},
+		},
+		{
+			desc:     "Subscribe to a topic with empty id",
+			topic:    topic,
+			clientID: "",
+			err:      mqtt_pubsub.ErrEmptyID,
+			handler:  handler{false, "", msgChan},
+		},
+	}
+	for _, tc := range cases {
+		err := pubsub.Subscribe(tc.clientID, tc.topic, tc.handler)
+		switch err {
+		case nil:
+			// use pubsub to subscribe to a topic, and then publish messages to that topic.
+			expectedMsg := messaging.Message{
+				Publisher: "clientID",
+				Channel:   channel,
+				Subtopic:  subtopic,
+				Payload:   data,
+			}
+
+			// publish message, and then receive it on message Channel
+			err := pubsub.Publish(topic, expectedMsg)
+			assert.Nil(t, err, fmt.Sprintf("%s: got unexpected error: %s\n", tc.desc, err))
+
+			receivedMsg := <-msgChan
+			assert.Equal(t, expectedMsg, receivedMsg, fmt.Sprintf("%s: expected %+v got %+v\n", tc.desc, expectedMsg, receivedMsg))
+		default:
+			assert.Equal(t, err, tc.err, fmt.Sprintf("%s: expected: %s, but got: %s", tc.desc, err, tc.err))
```

Comment:
Move this to L243, right after the `err := pubsub.Subscribe(tc.clientID, tc.topic, tc.handler)`. After that, use `if tc.err == nil {publish and so on....}` instead of a switch.
The same as in the previous test.